src/backend/storage/lmgr/lock.c

```diff
@@ -2059,8 +2060,12 @@ LockSetHoldTillEndXact(const LOCKTAG *locktag)
 		 * be NULL.
 		 * This code should be revised to make GDD work under fast path.
 		 */
-//		locallock->lock->holdTillEndXact = true;
+		if (locallock->lock)
+			locallock->lock->holdTillEndXact = true;
 	}
+
+	relid = locktag->locktag_field2;
+	setFPHoldTillEndXact(relid);
```

Comment:
@pivotal-ning-yu  I am not sure whether we could skip the for loop even if we have found some locks in fast path array.

Any ideas @guofengrichard

Reply:
@kainwen gdd logic is that once a relation has unbalanced ref counts we mark all its lock modes as unreleasable, for fast path could we assume that a lock is either tracked in fast path or in normal local hash but never both? If this assumption is true then we sure could skip the full loop. You could add some check or at least an assert here to verify this assumption.

Reply:
That's right. A relation lock would exist either in fast-pach or in shared lock table. You can refer to the logic `LockRelease()`.

Reply:
@guofengrichard 
Good news. I will improve the code.

Reply:
The logic in `LockRelease()` is first attempt fast release if the lock is eligible for the fast path. This attempt is not guaranteed to be always successful, because other backends may move the lock to hash table, or the lock is not entered in fast-path due to no available slot. 
I think that logic also applies in this function.